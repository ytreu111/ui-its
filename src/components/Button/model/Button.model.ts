import { ReactNode } from "react";
import CSS from "csstype";

import IPadding from "types/interface/IPadding";
import IMargin from "types/interface/IMargin";

export interface IButtonProps extends IPadding, IMargin{
  type?: ButtonType;
  size?: SizeType;
  disabled?: boolean;
  buttonStyle?: StyleType;
  icon?: ReactNode;
  iconSize?: IIconSize | number;
  iconColor?: CSS.Property.Color;
  iconMargin?: number;
  loading?: boolean;
  onClick?: (e: MouseEvent) => void;
  clickEffect?: boolean;
  rounded?: boolean;
  children?: ReactNode;
  fontSize?: CSS.Property.FontSize;
  width?: CSS.Property.Width;
}

interface IIconSize {
  width: CSS.Property.Width;
  height: CSS.Property.Height;
}

export enum ButtonTypeEnum {
  primary = 'primary',
  danger = 'danger',
  orange = 'orange',
  yellow = 'yellow',
  green = 'green',
  purple = 'purple',
  magenta = 'magenta',
  blue = 'blue',
  transparent = 'transparent',
}

export enum ButtonSizeEnum {
  large = 'large',
  default = 'default',
  small = 'small',
}

export enum ButtonStyleEnum {
  //TODO Временно убрано по просьбе дизайнеров, тк оно не используется в системе
  // stroke = 'stroke',
  default = 'default',
  translucent = 'translucent',
  transparent = 'transparent',
}

export type ButtonType = keyof typeof ButtonTypeEnum;

export type SizeType = keyof typeof ButtonSizeEnum;

export type StyleType = keyof typeof ButtonStyleEnum;