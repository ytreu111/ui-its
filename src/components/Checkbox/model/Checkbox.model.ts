import { ChangeEvent, ReactElement } from "react";

export interface ICheckboxProps {
  checked?: boolean;
  value?: any;
  moved?: boolean;
  disabled?: boolean;
  checkboxSize?: number;
  name?: string;
  iconRender?: (checked: boolean) => ReactElement;
  onChange?: (checked: boolean, value: any, event: ChangeEvent<HTMLInputElement>) => void;
}