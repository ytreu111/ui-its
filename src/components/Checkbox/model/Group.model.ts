export interface ICheckboxOption {
  label: string;
  value: any;
  disabled?: boolean;
}

export interface ICheckboxGroupProps {
  options?: Array<ICheckboxOption>;
  value?: Array<any>,
  onChange?: (values: Array<any>) => void;
}