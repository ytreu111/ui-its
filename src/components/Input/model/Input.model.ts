import { ReactNode } from 'react';

export interface InputProps {
  value?: string;
  disabled?: boolean;
  small?: boolean;
  placeholder?: string;
  editableHelp?: string;
  label?: string;
  icon?: ReactNode;
  error?: boolean;
  autoWidth?: boolean;
  onChange?: (value: string) => void;
  onBlur?: () => void;
  onFocus?: () => void;
  name?: string;
  width?: number | string;
  optional?: () => void;
  optionalIcon?: ReactNode;
}
