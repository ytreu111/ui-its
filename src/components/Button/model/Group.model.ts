import { StyleType } from './Button.model';
import { ReactNode } from 'react';

export interface IButtonGroupProps {
  value?: string;
  buttonGap?: number;
  selectButton?: StyleType,
  deselectButton?: StyleType,
  defaultIcon?: ReactNode,
  onChange?: (event: MouseEvent, value?: string) => void;
  defaultValue?: string;
}

type ButtonOptionsFields = 'selectButton' | 'deselectButton' | 'defaultIcon' | 'onChange';

export type ButtonOptions = Required<Pick<IButtonGroupProps, ButtonOptionsFields>>;