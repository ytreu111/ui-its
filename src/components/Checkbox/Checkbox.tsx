import { ChangeEvent, FC, MouseEvent, useCallback, useEffect, useMemo, useState } from 'react';
import { ThemeProvider } from 'styled-components';
import { CheckBoxIcon, ActiveCheckBoxIcon } from 'ui-its-icons';

import { ICheckboxProps } from './model/Checkbox.model';
import { Styled } from './style/Checkbox.styled';
import theme from 'styles/theme/theme';
import { CheckpointTestElemsEnum } from './__tests__/Checkpoint.test.model';

const {
  CheckboxWrap,
  CheckboxText,
  CheckboxIcon,
  CheckboxInput,
  CheckboxDragIcon,
} = Styled;

const Checkbox: FC<ICheckboxProps> = (
  {
    checked = false,
    value,
    checkboxSize,
    moved,
    disabled,
    name,
    iconRender,
    onChange,
    children,
  }) => {
  const [checkedInput, setCheckedInput] = useState<boolean>(false);

  useEffect(() => setCheckedInput(checked), [checked]);

  const checkboxIcon = useMemo(() => {
    if (iconRender) {
      return iconRender(checkedInput);
    } else {
      return checkedInput
        ? <ActiveCheckBoxIcon data-testid={CheckpointTestElemsEnum.activeIcon} />
        : <CheckBoxIcon data-testid={CheckpointTestElemsEnum.nonActiveIcon} />;
    }
  }, [iconRender, checkedInput]);

  const className = useMemo(() => {
    const className: Array<string> = [];

    if (disabled) className.push('disabled');

    return className.join(' ');
  }, [disabled]);

  const onChangeInput = useCallback((event: ChangeEvent<HTMLInputElement>) => {
    const target = event.currentTarget;
    const checked = target.checked;

    setCheckedInput(checked);

    if (onChange) onChange(checked, value, event);
  }, [onChange, value]);

  const onClickLabel = useCallback((event: MouseEvent<HTMLLabelElement>) => {
    if (moved || disabled) event.preventDefault();
  }, [disabled, moved]);

  return (
    <ThemeProvider theme={theme}>
      <CheckboxWrap onClick={onClickLabel} className={className}>
        <CheckboxIcon checkboxSize={checkboxSize}>{checkboxIcon}</CheckboxIcon>
        {moved && <CheckboxDragIcon data-testid={CheckpointTestElemsEnum.moveIcon} />}
        <CheckboxInput name={name} type="checkbox" onChange={onChangeInput} checked={checkedInput} />
        {children && <CheckboxText>{children}</CheckboxText>}
      </CheckboxWrap>
    </ThemeProvider>
  );
};

export default Checkbox;