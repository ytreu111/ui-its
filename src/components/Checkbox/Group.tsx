import React, { FC, ReactElement, ReactNode, useCallback, useEffect, useMemo, useState } from "react";
import { ThemeProvider } from "styled-components";

import { ICheckboxGroupProps, ICheckboxOption } from "./model/Group.model";
import { theme } from "styles";
import { Styled } from "./style/Group.styled";
import Checkbox from "./Checkbox";
import { ICheckboxProps } from "./model/Checkbox.model";

const {
  GroupWrap,
} = Styled;

const checkboxProps = (checkbox: ICheckboxOption, groupValue: Set<any>) => {
  const key: string = typeof checkbox.value === 'object' ? JSON.stringify(checkbox.value) : checkbox.value;
  const checked = groupValue.has(checkbox.value);

  return { key, checked };
}

const getCheckboxComponent = (checkbox: ReactElement<any, any>, groupValue: Set<any>, onChangeCheckbox: any) => {
  if (checkbox.type === Checkbox && checkbox.props?.value) {
    const { key, checked } = checkboxProps(checkbox.props, groupValue);

    return (
      <Checkbox {...checkbox.props} key={key} checked={checked} onChange={onChangeCheckbox}>
        {checkbox.props.children}
      </Checkbox>
    )
  }

  return null;
}

const childrenCheckbox = (children: ReactNode, groupValue: Set<any>, onChangeCheckbox: any, content?: Array<ReactNode>) => {
  let newContent = content ? [...content] : [];

  if (children && typeof children === 'object') {
    if (Array.isArray(children)) {
      for (const element of children) {
        if (element.type === React.Fragment) {
          newContent = childrenCheckbox(element.props.children, groupValue, onChangeCheckbox, newContent);
        } else {
          const newComponent = getCheckboxComponent(element, groupValue, onChangeCheckbox);
          newContent.push(newComponent);
        }
      }
    } else {
      if ('type' in children) {
        if (children.type === React.Fragment) {
          newContent = childrenCheckbox(children.props.children, groupValue, onChangeCheckbox, newContent);
        } else {
          const newComponent = getCheckboxComponent(children, groupValue, onChangeCheckbox);
          newContent.push(newComponent);
        }
      }
    }
  }

  return newContent;
}

const Group: FC<ICheckboxGroupProps> = (
  {
    options,
    value = [],
    onChange,
    children,
  }) => {
  const [groupValue, setGroupValue] = useState<Set<any>>();

  useEffect(() => setGroupValue(new Set(value)), [value]);

  const onChangeCheckbox = useCallback((checked: boolean, value: any) => {
    const newGroupValue = new Set(groupValue);

    if (newGroupValue.has(value)) {
      newGroupValue.delete(value);
    } else {
      newGroupValue.add(value);
    }

    setGroupValue(newGroupValue);
    if (onChange) onChange(Array.from(newGroupValue));
  }, [groupValue, onChange]);

  const content = useMemo(() => {
    let content: Array<ReactElement<ICheckboxProps, typeof Checkbox>> = [];

    if (options && !!options.length && groupValue) {
      content = options.map((el) => {
        const { key, checked } = checkboxProps(el, groupValue);

        return <Checkbox key={key} onChange={onChangeCheckbox} checked={checked} value={el.value}>{el.label}</Checkbox>
      })
    }

    return content;
  }, [groupValue, onChangeCheckbox, options]);

  const childrenContent = useMemo(() => {
    if (children && groupValue) {
      return childrenCheckbox(children, groupValue, onChangeCheckbox);
    }

    return null;
  }, [children, groupValue, onChangeCheckbox])

  return (
    <ThemeProvider theme={theme}>
      <GroupWrap>
        {content}
        {childrenContent}
      </GroupWrap>
    </ThemeProvider>
  );
}

export default Group;