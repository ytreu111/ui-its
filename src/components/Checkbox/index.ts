import { FC } from "react";

import InternalCheckbox from "./Checkbox";
import Group from "./Group";
import { ICheckboxProps } from "./model/Checkbox.model";

export interface CompoundedComponent extends FC<ICheckboxProps> {
  Group: typeof Group
}

const Checkbox = InternalCheckbox as CompoundedComponent;

Checkbox.Group = Group;

export type { ICheckboxProps } from './model/Checkbox.model'
export default Checkbox;