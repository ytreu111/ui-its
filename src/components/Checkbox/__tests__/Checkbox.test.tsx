import { render, screen, prettyDOM } from '@testing-library/react';
import Checkbox from '../Checkbox';
import userEvent from '@testing-library/user-event';
import { CheckpointTestElemsEnum } from './Checkpoint.test.model';
import { ActiveCheckBoxIcon, CheckBoxIcon } from 'ui-its-icons';

const onChange = jest.fn();

describe('Компонент Checkbox', () => {
  test('Наличие текста', () => {
    render(<Checkbox>checkbox</Checkbox>);
    expect(screen.getByText('checkbox')).toBeInTheDocument();
  });
  test('Наличие иконки при перемещении', () => {
    render(<Checkbox moved>checkbox</Checkbox>);
    expect(screen.getByTestId(CheckpointTestElemsEnum.moveIcon)).toBeInTheDocument();
  });
  test('Изменение иконок при клике', () => {
    render(<Checkbox onChange={onChange}>checkbox</Checkbox>);
    expect(screen.getByTestId(CheckpointTestElemsEnum.nonActiveIcon)).toBeInTheDocument();
    userEvent.click(screen.getByText('checkbox'));
    expect(screen.getByTestId(CheckpointTestElemsEnum.activeIcon)).toBeInTheDocument();
    expect(onChange).toBeCalledTimes(1);
  });
  test('Кастомная иконка', () => {
    render(<Checkbox
      iconRender={(checked) => checked
        ? <ActiveCheckBoxIcon data-testid={CheckpointTestElemsEnum.activeIcon} />
        : <CheckBoxIcon data-testid={CheckpointTestElemsEnum.nonActiveIcon} />}
    >
      checkbox
    </Checkbox>);

    expect(screen.getByTestId(CheckpointTestElemsEnum.nonActiveIcon)).toBeInTheDocument();
    userEvent.click(screen.getByText('checkbox'));
    expect(screen.getByTestId(CheckpointTestElemsEnum.activeIcon)).toBeInTheDocument();
  });
  test('disabled', () => {
    const { baseElement } = render(<Checkbox disabled>checkbox</Checkbox>);
    expect(prettyDOM(baseElement)).toMatchSnapshot();
    userEvent.click(screen.getByText('checkbox'));
    expect(screen.getByTestId(CheckpointTestElemsEnum.nonActiveIcon)).toBeInTheDocument();
  });
});