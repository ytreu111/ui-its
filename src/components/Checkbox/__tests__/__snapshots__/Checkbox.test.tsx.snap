// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`–ö–æ–º–ø–æ–Ω–µ–Ω—Ç Checkbox disabled 1`] = `
"[36m<body>[39m
  [36m<div>[39m
    [36m<label[39m
      [33mclass[39m=[32m\\"sc-eCImPb ldrcYY disabled\\"[39m
    [36m>[39m
      [36m<div[39m
        [33mclass[39m=[32m\\"sc-gsDKAQ kVIrCS\\"[39m
      [36m>[39m
        [36m<span[39m
          [33mclass[39m=[32m\\"anticon\\"[39m
          [33mdata-testid[39m=[32m\\"non-active-icon\\"[39m
          [33mrole[39m=[32m\\"img\\"[39m
        [36m>[39m
          [36m<svg[39m
            [33mfill[39m=[32m\\"currentColor\\"[39m
            [33mheight[39m=[32m\\"24\\"[39m
            [33mviewBox[39m=[32m\\"0 0 24 24\\"[39m
            [33mwidth[39m=[32m\\"24\\"[39m
            [33mxmlns[39m=[32m\\"http://www.w3.org/2000/svg\\"[39m
          [36m>[39m
            [36m<path[39m
              [33md[39m=[32m\\"M19 21H5a2 2 0 01-2-2V5a2 2 0 012-2h14a2 2 0 012 2v14a2 2 0 01-2 2zm0-2V5H5v14h14z\\"[39m
            [36m/>[39m
          [36m</svg>[39m
        [36m</span>[39m
      [36m</div>[39m
      [36m<input[39m
        [33mclass[39m=[32m\\"sc-hKwDye bmJdxt\\"[39m
        [33mtype[39m=[32m\\"checkbox\\"[39m
      [36m/>[39m
      [36m<span[39m
        [33mclass[39m=[32m\\"sc-bdvvtL jldYWA\\"[39m
      [36m>[39m
        [0mcheckbox[0m
      [36m</span>[39m
    [36m</label>[39m
  [36m</div>[39m
[36m</body>[39m"
`;
